app/Scripts/SpearLanceEnemy.gd

```diff
@@ -0,0 +1,162 @@
+extends KinematicBody2D
+
+
+var TargetPos: Vector2
+var Velocity := Vector2.ZERO
+var Target : Area2D
+var OverlappingTargetArray:= []
+var Dir := Vector2.ZERO
+var HasTarget := false
+onready var _animationPlayer = $AnimationPlayer
+onready var _attackTimer = $re_AttackTimer
+onready var _patrolTimer = $patrolTimer
+onready var _targetLockTimer = $TargetLockOnTimer
+var MoveTowardsTarget := false
+var isAttackingAnimation :=false
+func _ready():
+	_random_direction()
+	set_meta("Enemy", false)
+	isAttackingAnimation = false
+	Velocity = TargetPos - self.position
+	_animationPlayer.play("Idle")
+	Move()
+	PlayerStats.connect("Death", self, "TargetDead")
+func _on_Vision_area_entered(area): #if you dont have target, set a target
+	$Vision/CollisionShape2D.disabled = true
+	if HasTarget == false: 
+		
+		if area.get_parent().has_meta("Player") or area.get_parent().has_meta("Boid"):
+			Target = area
+			
+			MoveTowardsTarget = true
+			HasTarget = true
+			_patrolTimer.stop()
+			
+		
+func _physics_process(delta): #Target death
+	if Target != null && Target.get_parent().has_meta("Boid"):
+		if !BoidsGlobal.AllBoidsArray.has(Target.get_parent()):
```

Comment:
why do we need a global boids array? Isn't it enough to know whether the Target object is valid or not?

Reply:
for some reason, it doesn't understand when when queue_free() the boids so I have to do that to tell it the boid has died